File: src/pixel_picker.cpp
#include "pixel_picker.hpp"

PixelPicker::PixelPicker(double shrink_height_=0.75, int dot_radius_=5){
    shrink_height = shrink_height_;
    dot_radius = dot_radius_;
    point.x = -1;
    point.y = -1;
}


cv::Point PixelPicker::pickOne(cv::Mat img){
    // First create the image with alpha channel
    cv::cvtColor(img, rgba, cv::COLOR_RGB2RGBA);
    std::vector<cv::Mat> channels(4);
    // Then assign the mask to the last channel of the image
    cv::split(rgba, channels);
    // cv::Mat alpha(img.rows, img.cols, CV_8UC1, cv::Scalar(128));
    channels[3] = cv::Mat(img.rows, img.cols, CV_8UC1, cv::Scalar(128));
    // modify channel// then merge
    cv::merge(channels, rgba);

    cv::namedWindow("img");

    rgba.copyTo(res);
    resshow = ResizeWithAspectRatio(res, -1, int(res.rows*shrink_height));
    cv::imshow("img", resshow);
    cv::setMouseCallback("img", PixelPicker::mouser, this);
    cv::waitKey(0);
    cv::destroyAllWindows();

    return point;
}

// mutators
void PixelPicker::set_shrink_height(double shrink_height_){
    shrink_height = shrink_height_;
}

void PixelPicker::set_dot_radius(int dot_radius_){
    dot_radius = dot_radius_;
}

void PixelPicker::mouser(int event, int x, int y, int flags, void* this_) {
  static_cast<PixelPicker*>(this_)->onMouse(event, x, y);
}

void PixelPicker::onMouse(int evt, int x, int y) {
    rgba.copyTo(res);
    x = int(x/shrink_height);
    y = int(y/shrink_height);
    if(evt == cv::EVENT_LBUTTONDOWN) {
        point.x = x;
        point.y = y;
        cv::line(res, cv::Point(x, 0), cv::Point(x, res.rows), (0,0,255,128), 2);
        cv::line(res, cv::Point(0, y), cv::Point(res.cols, y), (0,0,255,128), 2);
    }
    // cnd = redDot[:, :, 3] > 0;
    // res[cnd] = redDot[cnd];
    resshow = ResizeWithAspectRatio(res, -1, int(res.rows*shrink_height));

    cv::imshow("img", resshow);
}

cv::Mat PixelPicker::ResizeWithAspectRatio(cv::Mat img, int width=-1, int height=-1){
    cv::Mat out;
    int h = img.rows;
    int w = img.cols;
    double r;

    if (width==-1 && height==-1){
        return img;
    }
    if (width==-1){
        r = 1.0*height/h;
        // dim = cv::Size(int(w * r), height);
        cv::resize(img, out, cv::Size(int(w * r), height), cv::INTER_AREA);
        return out;
    }
    else{
        r = 1.0*width/w;
        // dim = cv::Size(width, int(h * r));
        cv::resize(img, out, cv::Size(width, int(h * r)), cv::INTER_AREA);
        return out;
    }
}

File: src/color_picker/pixel_picker.py
import cv2
import numpy as np

shrinkHeight = 1.0
dotRadius = 5

def PointPickerTarCur(img, height_scale = shrinkHeight, r=dotRadius):

    global points, bluDot, redDot, greDot, res, rgba
    points = [[-1, -1], [-1, -1], [-1, -1]]
    
    # First create the image with alpha channel
    rgba = cv2.cvtColor(img, cv2.COLOR_RGB2RGBA)

    # Then assign the mask to the last channel of the image
    
    alphaChannel = np.full((img.shape[0], img.shape[1]), 128)
    rgba[:, :, 3] = alphaChannel
    #img = np.dstack((img, alphaChannel))
    
    bluDot = np.zeros((img.shape[0], img.shape[1], 4), dtype=np.uint8)
    redDot = np.zeros((img.shape[0], img.shape[1], 4), dtype=np.uint8)  
    greDot = np.zeros((img.shape[0], img.shape[1], 4), dtype=np.uint8)  
    
    def mouse_click1(event, x, y, 
                flags, param):
        
        actualX = int(x/height_scale)
        actualY = int(y/height_scale)

        global points, redDot, res, rgba
        if event == cv2.EVENT_LBUTTONDOWN:

            points[0] = [actualX,actualY]
            
            redDot = np.zeros((img.shape[0], img.shape[1], 4), dtype=np.uint8)  
            cv2.line(redDot, (actualX, 0), (actualX, redDot.shape[0]), (0,0,255,128), 2)
            cv2.line(redDot, (0, actualY), (redDot.shape[1], actualY), (0,0,255,128), 2)
        
        res = rgba.copy()
        cnd = redDot[:, :, 3] > 0
        res[cnd] = redDot[cnd]
        resshow = ResizeWithAspectRatio(res, height=int(res.shape[0]*height_scale))

        cv2.imshow('image', resshow)
        
    def mouse_click2(event, x, y, 
                flags, param):
        
        actualX = int(x/height_scale)
        actualY = int(y/height_scale)

        global points, bluDot, res, rgba
        if event == cv2.EVENT_LBUTTONDOWN:
            
            points[1] = [actualX,actualY]
            bluDot = np.zeros((img.shape[0], img.shape[1], 4), dtype=np.uint8)
            cv2.line(bluDot, (actualX, 0), (actualX, bluDot.shape[0]), (255,0,0,128), 2)
            cv2.line(bluDot, (0, actualY), (bluDot.shape[1], actualY), (255,0,0,128), 2)
        
        res = rgba.copy()
        cnd = bluDot[:, :, 3] > 0
        res[cnd] = bluDot[cnd]
        resshow = ResizeWithAspectRatio(res, height=int(res.shape[0]*height_scale))

        cv2.imshow('image', resshow)
    
    def mouse_click3(event, x, y, 
                flags, param):
        
        actualX = int(x/height_scale)
        actualY = int(y/height_scale)

        global points, greDot, res, rgba
        if event == cv2.EVENT_LBUTTONDOWN:
            
            points[2] = [actualX,actualY]
            greDot = np.zeros((img.shape[0], img.shape[1], 4), dtype=np.uint8)
            cv2.line(greDot, (actualX, 0), (actualX, greDot.shape[0]), (0,255,0,128), 2)
            cv2.line(greDot, (0, actualY), (greDot.shape[1], actualY), (0,255,0,128), 2)

        res = rgba.copy()
        cnd = greDot[:, :, 3] > 0
        res[cnd] = greDot[cnd]
        resshow = ResizeWithAspectRatio(res, height=int(res.shape[0]*height_scale))

        cv2.imshow('image', resshow)
        
    
    res = rgba.copy()
    resshow = ResizeWithAspectRatio(res, height=int(res.shape[0]*height_scale))
    cv2.imshow('image', resshow)   
    cv2.setMouseCallback('image', mouse_click1)
    cv2.waitKey(0)
    cv2.destroyAllWindows()

    res = rgba.copy()
    resshow = ResizeWithAspectRatio(res, height=int(res.shape[0]*height_scale))
    cv2.imshow('image', resshow)   
    cv2.setMouseCallback('image', mouse_click2)
    cv2.waitKey(0)
    cv2.destroyAllWindows()
    
    res = rgba.copy()
    resshow = ResizeWithAspectRatio(res, height=int(res.shape[0]*height_scale))
    cv2.imshow('image', resshow)   
    cv2.setMouseCallback('image', mouse_click3)
    cv2.waitKey(0)
    cv2.destroyAllWindows()

    return points

def PointPickerOne(img, height_scale = shrinkHeight, r=dotRadius):

    global points, bluDot, redDot, greDot, res, rgba
    points = [-1, -1]
    
    # First create the image with alpha channel
    rgba = cv2.cvtColor(img, cv2.COLOR_RGB2RGBA)

    # Then assign the mask to the last channel of the image
    
    alphaChannel = np.full((img.shape[0], img.shape[1]), 128)
    rgba[:, :, 3] = alphaChannel
    #img = np.dstack((img, alphaChannel))
    
    redDot = np.zeros((img.shape[0], img.shape[1], 4), dtype=np.uint8)  
    
    def mouse_click(event, x, y, 
                flags, param):
        
        actualX = int(x/height_scale)
        actualY = int(y/height_scale)

        global points, redDot, res, rgba
        if event == cv2.EVENT_LBUTTONDOWN:

            points = [actualX,actualY]
            
            redDot = np.zeros((img.shape[0], img.shape[1], 4), dtype=np.uint8)  
            cv2.line(redDot, (actualX, 0), (actualX, redDot.shape[0]), (0,0,255,128), 2)
            cv2.line(redDot, (0, actualY), (redDot.shape[1], actualY), (0,0,255,128), 2)
        
        res = rgba.copy()
        cnd = redDot[:, :, 3] > 0
        res[cnd] = redDot[cnd]
        resshow = ResizeWithAspectRatio(res, height=int(res.shape[0]*height_scale))

        cv2.imshow('image', resshow)     
    
    res = rgba.copy()
    resshow = ResizeWithAspectRatio(res, height=int(res.shape[0]*height_scale))
    cv2.imshow('image', resshow)   
    cv2.setMouseCallback('image', mouse_click)
    cv2.waitKey(0)
    cv2.destroyAllWindows()

    return points

def ResizeWithAspectRatio(image, width=None, height=None, inter=cv2.INTER_AREA):
    dim = None
    (h, w) = image.shape[:2]

    if width is None and height is None:
        return image
    if width is None:
        r = height / float(h)
        dim = (int(w * r), height)
    else:
        r = width / float(w)
        dim = (width, int(h * r))

    return cv2.resize(image, dim, interpolation=inter)
